write-fonts/src/tables/glyf.rs

```diff
@@ -86,29 +88,291 @@ pub enum BadKurbo {
+        for (i, (p0, p1)) in p0s.into_iter().zip(p1s.into_iter()).enumerate() {
+            self.0[i].push(ContourPoint::off_curve(p0));
+            self.0[i].push(ContourPoint::on_curve(p1));
+        }
+    }
+
+    /// The total number of points in each interpolatable contour
+    fn num_points(&self) -> usize {
+        let n = self.0[0].len();
+        assert!(self.0.iter().all(|c| c.len() == n));
+        n
+    }
+
+    /// The first point in each contour
+    fn first(&self) -> Vec<&ContourPoint> {
+        self.0.iter().map(|v| v.first().unwrap()).collect()
+    }
+
+    /// The last point in each contour
+    fn last(&self) -> Vec<&ContourPoint> {
+        self.0.iter().map(|v| v.last().unwrap()).collect()
+    }
+
+    /// Remove the last point from each contour
+    fn remove_last(&mut self) {
+        self.0.iter_mut().for_each(|c| {
+            c.pop().unwrap();
+        });
+    }
+
+    /// Compute the intersection of all implied on-curve points and return their indices
+    fn impliable_oncurve_points(&self) -> HashSet<usize> {
+        let epsilon = f32::EPSILON as f64; // should we make it a parameter?
+        let mut result = HashSet::new();
+        if let Some((first, others)) = self.0.split_first() {
+            result = impliable_oncurve_points(first, epsilon);
+            for points in others {
+                result = result
+                    .intersection(&impliable_oncurve_points(points, epsilon))
+                    .copied()
+                    .collect();
+            }
+        }
+        result
+    }
+
+    /// Build the contours, dropping any on-curve points that can be implied in all contours
+    fn build(self) -> Vec<Contour> {
+        let drop = self.impliable_oncurve_points();
+        // Omit implied on-curve points when building the contours
+        let mut contours = Vec::with_capacity(self.len());
+        for points in self.0 {
+            contours.push(Contour(
+                points
+                    .into_iter()
+                    .enumerate()
+                    .filter_map(|(i, pt)| {
+                        if drop.contains(&i) {
+                            None
+                        } else {
+                            Some(CurvePoint::from(pt))
+                        }
+                    })
+                    .collect(),
+            ))
+        }
+        contours
+    }
 }
 
-/// A helper trait for converting other point types to open-type compatible reprs
-pub trait OtPoint {
-    fn get(self) -> (i16, i16);
+/// Indices of on-curve points that are at equal distance between two off-curve points
+fn impliable_oncurve_points(points: &[ContourPoint], tolerance: f64) -> HashSet<usize> {
+    let mut result = HashSet::new();
+    for (i, p1) in points.iter().enumerate() {
+        if !p1.on_curve {
+            continue;
+        }
+        let prv = if i > 0 { i - 1 } else { points.len() - 1 };
+        let nxt = if i < points.len() - 1 { i + 1 } else { 0 };
```

Comment:
I suggest you implement a wrapping read function and use it here, e.g. prev = wrapping_read(i-1), etc

Reply:
#377 covers this